### PR TITLE
Assignment7+8: Oscar & David

### PR DESCRIPTION
**Pull request by Oscar-Ruelas**
Collab w/ @dav1dmoore

### CODE REVIEW

**Comment by JacobLucas02** [comment]
## Instructions

**Throughout, consider using the `System.Linq.Enumerable` methods `Zip`, `Count`, `Sort` and `Contains` methods for testing collections.**. (Preferably avoid using `Microsoft.VisualStudio.TestTools.UnitTesting.CollectionAssert` although that might be easier, to get a firmer grasp on additional LINQ API.)

1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ✔

   - Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
   - Using LINQ, skip the first row in the `People.csv`. ✔
   - Be sure to appropriately handle resource (`IDisposable`) items correctly if applicable (and it may not be depending on how you implement it). ✔

2. Implement `IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()` to return a **sorted**, **unique** list of states. ✔

   - Use `ISampleData.CsvRows` for your data source. ✔
   - Don't forget the list should be unique. ✔
   - Sort the list alphabetically. ✔
   - Include a test that leverages a hardcoded list of addresses. ✔
   - Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔

3. Implement `ISampleData.GetAggregateSortedListOfStatesUsingCsvRows()` to return a `string` that contains a **unique**, comma separated list of states. ✔

   - Use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows()` for your data source. ✔
   - Consider "selecting" only the states and calling `ToArray()` to retrieve an array of all the state names. ✔
   - Given the array, consider using `string.Join` to combine the list into a single string. ✔

4. Implement the `ISampleData.People` property to return all the items in `People.csv` as `Person` objects ✔

   - Use `ISampleData.CsvRows` as the source of the data. ✔
   - Sort the list by State, City, and Zip. ✔
   - Be sure that `Person.Address` is also populated. ✔
   - Adding null validation to all the `Person` and `Address` properties is **optional**.
   - Consider using `ISampleData.CsvRows` in your test to verify your results. ✔

5. Implement `ISampleDate.FilterByEmailAddress(Predicate<string> filter)` to return a list of names where the email address matches the `filter`. ✔

   - Use `ISampleData.People` for your data source. ✔

6. Implement `ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)` to return a `string` that contains a **unique**, comma-separated list of states. ✔

   - Use the `people` parameter from `ISampleData.People` property for your data source. ✔
   - At a minimum, use the `System.Linq.Enumerable.`Aggregate` LINQ method to create your result. ✔
   - Don't forget the list should be unique. ✔
   - It is recommended that, at a minimum, you use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` to validate your result.

7. Given the implementation of `Node` in Assignment5

- Implement `IEnumerable<T>` to return all the items in the "circle" of items. ✔
- Add an `IEnumberable<T> ChildItems(int maximum)` method to `Node` that returns the remaining items with a maximum number of items returned less than `maximum`.  

## Extra Credit

- Implement the homework using async/await and multi-threading by defining a new `SampleDataAsync` class that implements `IAsyncSampleData`). Refactor your `SampleData` and `SampleDataAsync` classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file.
- Place all shared project items into a `Directory.Build.targets` file.
- Ensure nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ✔
- Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, consider using `Assert.AreEqual<T>()` (the generic version)  ❌
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

Overall, good job on this assignment! Everything is mostly implemented. The only thing I didn't see was the Assert.AreEqual() and extra credit.

**Comment by Oscar-Ruelas** [comment]
@JacobLucas02 I noticed that you marked For this assignment, consider using Assert.AreEqual<T>() (the generic version) as not satisfied, but we are using xUnit, and in xUnit, Assert.Equal provides the same functionality in a simpler and more intuitive way. While MSTest requires the explicit generic method Assert.AreEqual<T>() for type-safe comparisons, xUnit achieves this seamlessly with automatic type inference. Since xUnit’s Assert.Equal aligns with the intent and functionality of Assert.AreEqual<T>(), this requirement is fully satisfied without needing a change in frameworks or added complexity. Mark had even told us that if were using xunit we can ignore the requirement of  Assert.AreEqual<T>() 

that being said can you please update your review?  Thank you

File: Assignment/SampleData.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+{
+    // 1.
+    /*
+        Property to read lines from the CSV file "People.csv". The first line (header) is skipped.
+        This uses File.ReadLines to avoid loading the entire file into memory, which is useful for large files.
+    */
+    public IEnumerable<string> CsvRows { get; set; } = File.ReadLines("People.csv").Skip(1);
+
+    // 2.
+    public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
```
Comment (dpalmer10) [inline Assignment/SampleData.cs:18]: Doesn't affect the functionality at all, but you could potentially change this to a lambda expression,  and avoid using the "return" statement. 
```suggestion
    public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() => CsvRows.Select(row => row.Split(",")[6].Trim())
    .Distinct()
    .Orderby(state => state);
```

**Review by dpalmer10 — COMMENTED** [review]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
        Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
        Using LINQ, skip the first row in the People.csv. ✔
        Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔

    Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
        Use ISampleData.CsvRows for your data source. ✔
        Don't forget the list should be unique. ✔
        Sort the list alphabetically. ✔
        Include a test that leverages a hardcoded list of addresses. ✔
        Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔

    Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
        Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
        Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
        Given the array, consider using string.Join to combine the list into a single string. ✔

    Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
        Use ISampleData.CsvRows as the source of the data. ✔
        Sort the list by State, City, and Zip. ✔
        Be sure that Person.Address is also populated. ✔
        Adding null validation to all the Person and Address properties is optional.
        Consider using ISampleData.CsvRows in your test to verify your results. ✔

    Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔
        Use ISampleData.People for your data source. ✔

    Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔
        Use the people parameter from ISampleData.People property for your data source. ✔
        At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
        Don't forget the list should be unique. ✔
        It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.

    Given the implementation of Node in Assignment5

    Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
    Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.

Extra Credit

    Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌✔

Fundamentals

    Place all shared project properties into a Directory.Build.Props file.
    Place all shared project items into a Directory.Build.targets file.
    Ensure nullable reference types is enabled ✔
    Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
    Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
    and enabled .NET analyzers for both projects ✔
    For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
    All of the above should be unit tested ✔
    Choose simplicity over complexity ✔


Looks great, good work!

File: Assignment/SampleData.cs (hunk 1 above)
Comment (Oscar-Ruelas) [reply Assignment/SampleData.cs:18]: Ok I'm about to work on this

Comment (Oscar-Ruelas) [reply Assignment/SampleData.cs]: Ok the code has been updated

File: Assignment/SampleData.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+    // 4.
+    public IEnumerable<IPerson> People
+    {
+        get
+        {
+            return CsvRows
+                .Select(row =>
+                {
+                    var columns = row.Split(',');
```
Comment (quattro004) [inline Assignment/SampleData.cs:50]: I noticed this Split() call above with a comma delimiter as well. This class is assuming that the CSV file will contain comma delimited values and indeed the C in CSV stands for comma but experience has shown me that people use other delimiters as well, an enhancement to this class would be to allow the consumer to specify the delimiter.

Comment (Oscar-Ruelas) [reply Assignment/SampleData.cs:50]: will this delimiter be requested from consumer itself? like using readline right?

**Review by quattro004 — APPROVED** [review]
Looks good :shipit:

File: Assignment/SampleData.cs (hunk 2 above)
Comment (quattro004) [reply Assignment/SampleData.cs:50]: You could add a property which is defaulted to a "," or maybe include a string delimiter parameter to the constructor. Then in the split code you could use this variable like: row.Split(SplitDelimiter);

File: Assignment/SampleData.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+                    var address = new Address(
+                        columns[4].Trim(),
```
Comment (BenjaminMichaelis) [inline Assignment/SampleData.cs:53]: Nit: giving a value to these seemingly arbitrary numbers for columns to give a understandable name (like a static or enum or similar) could be nice. ex: `Column.EmailAddress` :)

File: Assignment/SampleData.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            .Select(person => (person.FirstName, person.LastName));
+    }
+
+    // 6.
+    public string GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)
+    {
+        if (people == null)
+        {
+            throw new ArgumentNullException(nameof(people), "The people collection cannot be null.");
+        }
```
Comment (BenjaminMichaelis) [inline Assignment/SampleData.cs:91]: nit: `ArgumentNullException.ThrowIfNull(people);`

**Pull request closed without merging**